elastic/kibana
# Fixes Spaces Menu Keyboard and Screen Reader Navigation

Fixes #130898

## Summary

Refactor of the Spaces popover implementation using the [EuiSelectable component](https://elastic.github.io/eui/#/forms/selectable). This component handles all of the focus, tabbing, and search behaviors innately - resolving previous navigation issues, and keeping a consistent UX in Kibana.

A relevant demonstration of the component can be found [here](https://elastic.github.io/eui/#/layout/header).

### Minor Behavior Change
When the user selects the already active space in the menu without any options (shift, ctrl/cmd, middle click), they will not be redirected to the home screen and instead remain on the current screen. The previous implementation would perform a full page redirect to the home screen, even if the user selected the space that is already active.

## Testing Non-Searchable Mode:
1. Create up to 6 spaces, for a maximum total of 7 including the default space
2. Verify the spaces navigation button avatar reflects the current active space
3. Click the spaces navigation button, verify a popover with heading "Your spaces" appears displaying all spaces with avatars on the left and a "Manage spaces" button at the bottom.
4. Verify the active space is checked.
5. Verify that the tab key toggles focus between the popover frame, the spaces list, and the manage spaces button.
6. Verify that that the up and down arrow keys can be used to navigate the spaces list when the spaces list is in focus.
7. Verify that selection will wrap when the arrow keys are used at the list limits (last option to first option, first option to last).
8. Verify that pressing enter when the manage spaces button is in focus causes the manage space app to load.
9. Verify that pressing enter when the spaces list is in focus and a non-active space is selected causes the select space to become the active space (full page refresh).
10. Verify that holding shift and pressing enter when the spaces list is in focus and a non-active space is selected causes the select space to open in a new window.
11. Verify that holding command (control on Windows) and pressing enter when the spaces list is in focus and a non-active space is selected causes the select space to open in a new tab.
12. Verify that using the mouse to click on a non-active space in the spaces list causes the select space to become the active space (full page refresh).
13. Verify that holding shift and clicking on a non-active space in the spaces causes that space to open in a new window.
14. Verify that holding command (control on Windows) and clicking on a non-active space in the spaces list causes that space to open in a new tab.
15. Verify the spaces navigation popover can be closes by clicking outside the popover, pressing escape, or by clicking the spaces navigation button when the popover is already open.

## Testing Searchable Mode:
1. Create at least 7 spaces, for a minimum total of 8 including the default space
2. Verify the spaces navigation button avatar reflects the current active space
3. Click the spaces navigation button, verify a popover with a search input appears with placeholder text "Find a space", verify the popover also contains a list of all spaces with avatars on the left and a "Manage spaces" button at the bottom.
4. Verify the active space is checked.
5. Verify that the tab key toggles focus between the popover frame, the spaces search input, and the manage spaces button.
6. Verify that that the up and down arrow keys can be used to navigate the spaces list when the spaces search field is in focus.
7. Verify that selection will wrap when the arrow keys are used at the list limits (last option to first option, first option to last).
8. Verify that pressing enter when the manage spaces button is in focus causes the manage space app to load.
9. Verify that typing text into the spaces search field causes the options in the spaces list to be filtered based on the input.
10. Verify that pressing enter when the spaces search field is in focus and a non-active space is selected causes the select space to become the active space (full page refresh).
11. Verify that holding shift and pressing enter when the spaces search field is in focus and a non-active space is selected causes the select space to open in a new window.
12. Verify that holding command (control on Windows) and pressing enter when the spaces search list is in focus and a non-active space is selected causes the select space to open in a new tab.
13. Verify that using the mouse to click on a non-active space in the spaces list causes the select space to become the active space (full page refresh).
14. Verify that holding shift and clicking on a non-active space in the spaces causes that space to open in a new window.
15. Verify that holding command (control on Windows) and clicking on a non-active space in the spaces list causes that space to open in a new tab.
17. Verify the spaces navigation popover can be closes by clicking outside the popover, pressing escape, or by clicking the spaces navigation button when the popover is already open.

## Testing The Screen Reader:
1. Enable a screen reader
2. Verify that when the spaces navigation header button is in focus that the reader states the currently selected space and the purpose of the button.
3. Open the spaces navigation button with the return key.
4. Verify the popover opens.
5. Use the keyboard to navigate into the popover.
6. Verify the screen reader states that you're in a text entry field of a list when the search field is in focus.
8. When traversing the spaces list using the up and down keys, verify the screen reader state the selected option in the list.
9. When using the search field, verify the screen reader states the entered text and number of options.
10. Verify the screen reader states "manage spaces button" when the manage space button is in focus.
11. Verify the screen reader still works appropriately when the search field is not present (7 or fewer spaces).

Note: in my testing I found Safari to exhibit undesirable screen-reader and tabbing behavior both with the new Spaces navigation implementation and the above linked EUI demonstration.
